grpclib/protocol.py

```diff
@@ -223,13 +230,88 @@ def flush(self) -> None:
+        if self._config._keepalive_time is not None:
+            self._ping_handle = asyncio.get_event_loop().call_later(
+                self._config._keepalive_time,
+                self._ping
+            )
+
     def close(self) -> None:
         if hasattr(self, '_transport'):
             self._transport.close()
             # remove cyclic references to improve memory usage
             del self._transport
             if hasattr(self._connection, '_frame_dispatch_table'):
                 del self._connection._frame_dispatch_table
+        if self._ping_handle is not None:
+            self._ping_handle.cancel()
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+
+    def _ping(self) -> None:
+        data = struct.pack('!Q', int(time.monotonic() * 10 ** 6))
+        is_need_send = True
+
+        if not self._config._keepalive_permit_without_calls:
+            count_of_connection = any(
+                s.open for s in self._connection.streams.values()
+            )
+            if count_of_connection == 0:
+                is_need_send = False
+
+        _keepalive_time = cast(float, self._config._keepalive_time)
+        current_time = time.monotonic()
+        if current_time - self._last_received() < _keepalive_time:
+            is_need_send = False
+
+        if self.last_ping_sent is not None and \
+                current_time - self.last_ping_sent < \
+                self._config._http2_min_sent_ping_interval_without_data:
+            is_need_send = False
+
+        if self._config._http2_max_pings_without_data != 0 and \
+                self.ping_count_in_sequence >= \
+                self._config._http2_max_pings_without_data:
+            is_need_send = False
+
+        if is_need_send:
+            logging.error(f'send ping {self._config._keepalive_timeout}')
+            self._connection.ping(data)
+            self.flush()
+            self.last_ping_sent = time.monotonic()
+            self.ping_count_in_sequence += 1
+            if self._close_by_ping_handler is None:
+                self._close_by_ping_handler = asyncio.get_event_loop().\
+                    call_later(
+                        self._config._keepalive_timeout,
+                        self.close
+                    )
+        self._ping_handle = asyncio.get_event_loop().call_later(
+            _keepalive_time,
+            self._ping
+        )
+
+    def data_process(self) -> None:
+        self.ping_count_in_sequence = 0
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+            self._close_by_ping_handler = None
+
+    def ping_process(self) -> None:
+        if self._config._keepalive_time is None:
+            return
+        if self._close_by_ping_handler is not None:
+            self._close_by_ping_handler.cancel()
+        self._close_by_ping_handler = asyncio.get_event_loop().call_later(
```

Comment:
This is a duplicate. I think that this timer should be created only after we send a PING frame and only `_ping` method is responsible to do this.

Reply:
I think that it depends on what happens when we used the maximum number of pings. This option assumes that the connection closes after the last ping.

Reply:
Since we have a periodically firing `_ping` method, we can place this logic there.

Reply:
> This option assumes that the connection closes after the last ping.

This behaviour is not documented by the gRPC authors, or can you point me to it? I find that it is useful to close idle connections, but maybe this feature is out of scope of keepalive feature and should be implemented in an additional PR?

Reply:
removed